### PR TITLE
[v14] Fix multiple upgrader issues

### PR DESCRIPTION
**Pull request by Migaroez**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Fixed issues
- [x] An error is thrown during unattended upgrade because of changes to the preview hub that could no longer deal with full cache refreshes (no key is present)
- [x] When the migrations changes the user in such a way that the link between the already issued tokens and the user store object is broken, authorization will throw server errors => a revoke token mechanism is introduced that can be called by the upgrader when a migration deems it necessary.

### Open related issues
- After login to authorize the upgrade, the user is redirected to the back office instead of the upgrade screen. This results in the back office being in a bad state due to migrations not having ran yet. This can be circumvented by refreshing the page.

### Testing
- create a new v13 latest database with 1 custom datatype/contenttype/document
- use this database to run the upgrader both unattended and attended. The only issue that should remain is having to refresh the page after login as described above.

### CODE REVIEW

File: src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs [hunk 1]
```diff
@@ -135,6 +144,12 @@ public ExecutedMigrationPlan ExecutePlan(MigrationPlan plan, string fromState)
             RebuildCache();
         }
 
+        // If any completed migration requires us to sign out the user we'll do that.
+        if (_invalidateBackofficeUserAccess)
+        {
+            RevokeBackofficeTokens().GetAwaiter().GetResult(); // should async all the way up at some point
```
Comment (Migaroez) [inline src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs:150]: I myself am not happy about this.
Had a quick look up the chain and it seems doable to change everything to async but I might have missed a branch where it isn't

File: src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs [hunk 2]
```diff
@@ -320,11 +335,43 @@ private void RunMigration(Type migrationType, MigrationContext context)
...
         _publishedSnapshotService.RebuildAll();
         _distributedCache.RefreshAllPublishedSnapshot();
     }
+
+    private async Task RevokeBackofficeTokens()
+    {
+        using IServiceScope scope = _serviceScopeFactory.CreateScope();
+
+        IOpenIddictApplicationManager openIddictApplicationManager = scope.ServiceProvider.GetRequiredService<IOpenIddictApplicationManager>();
+        var backOfficeClient = await openIddictApplicationManager.FindByClientIdAsync(Constants.OAuthClientIds.BackOffice);
```
Comment (Migaroez) [inline src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs:356]: Not happy that we only invalidate our own client, the only way to make this extendable is to use the options pattern and loop over all the clients configured that way.
Might be overkill 🤷

Comment (kjac) [reply src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs:356]: I think it's overkill. Users are tied to the core clients. We _could_ consider handling the Swagger and Postman clients too, but really... those are just for dev, let's not bother with them now.

File: src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs (hunk 1 above)
Comment (Migaroez) [reply src/Umbraco.Infrastructure/Migrations/MigrationPlanExecutor.cs:150]: Glad you agree with my stomach turning over thinking about all the signature changes 😄

**Comment by Migaroez** [comment]
Dev merge issues and broken test fixes are incoming

**Comment by kjac** [comment]
Uh. I think we may have been fixing the same issue with the preview hub 🤦 https://github.com/umbraco/Umbraco-CMS/pull/16338 has been merged in and solves the same issue.

**Comment by Migaroez** [comment]
> Uh. I think we may have been fixing the same issue with the preview hub 🤦 #16338 has been merged in and solves the same issue.

Yup, merged your correction in on my machine already, will be pushed shortly

**Pull request merged**